# PES University, Bangalore
Established under Karnataka Act No. 16 of 2013

## UE22AM343AB4 - Advanced Data Analytics

Designed by Sathwik HJ

### Student Details
- Name : **Student Name**
- SRN : **SRN**

## Market Basket Analysis
    + Apriori Algorithm
    + Association Rule mining

### Prerequisites
- Revise the following concepts
    - Apriori Algorithm
        - Suport
        - Confidence
        - Lift
- Install the following software
    - pandas
    - apyori

### Problems

1. Problem 1: Preprocessing
2. Problem 2: Item set detection and analaysis
3. Problem 3: Association rule minning
- Note: Make sure to include justifications for each of your approaches and also provide a brief observations of the results obtained

### Context
Welcome to "Shocking Electronics", a budding Electronic Store. As a data analyst working with "Shocking Electronics" , your mission is to uncover meaningful patterns within customer transactions to enhance their shopping experience and help us compete with our competitor "Circuit Circus Electronics".

In [1]:
!wget https://raw.githubusercontent.com/MBUYt0n/ada/main/ADA_Worksheet_C/Electronics.csv

--2024-09-20 04:12:00--  https://raw.githubusercontent.com/MBUYt0n/ada/main/ADA_Worksheet_C/Electronics.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9593 (9.4K) [text/plain]
Saving to: ‘Electronics.csv’

Electronics.csv     100%[===================>]   9.37K  --.-KB/s    in 0s      

2024-09-20 04:12:00 (51.5 MB/s) - ‘Electronics.csv’ saved [9593/9593]



### About the dataset
- "Electronics.csv"
- Each record of the dataset represents a transaction made by a customer at "Shocking Electronics store".
- The transaction contains the items bought in that transaction

In [2]:
# Install apyori package
# Use % or ! based on the environment to install
%pip install apyori
%pip install mlxtend
import apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5953 sha256=062e912eb7c82dfe55ef5bf94539411c40d9c5c6c1e9d5ec9c8234be61324f5c
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


apyori documentation: https://pypi.org/project/apyori/ (Refer API Usage)

In [3]:
import pandas as pd
from apyori import apriori

### Problem 1 - Preprocessing
**Load the transactions data** from the provided csv file. **Transform the data** to a suitable format (Hint: **List of lists**[internal list contains the items of the transaction]). **Make sure to clean the data** (Hint: NA values).


In [4]:
df = pd.read_csv('Electronics.csv', names=["item1", "item2", "item3", "item4", "item5", "item6"], index_col=False)
df

item1                item2                item3  \
0          smartphone              earbuds     screen protector   
1              laptop                mouse             keyboard   
2          smartwatch      fitness tracker     wireless charger   
3                  TV             soundbar           HDMI cable   
4              tablet               stylus          tablet case   
..                ...                  ...                  ...   
212  desktop computer  mechanical keyboard         gaming mouse   
213        smartwatch      fitness tracker        charging dock   
214            laptop               webcam  external hard drive   
215            tablet     screen protector               stylus   
216  desktop computer       gaming monitor         RGB keyboard   

                   item4                item5       item6  
0             power bank            USB cable  phone case  
1           laptop stand  external hard drive         NaN  
2                    NaN                  NaN         NaN  
3       streaming device                  NaN         NaN  
4         screen cleaner                  NaN         NaN  
..                   ...                  ...         ...  
212                  NaN                  NaN         NaN  
213  wireless headphones                  NaN         NaN  
214              USB hub                  NaN         NaN  
215     portable charger                  NaN         NaN  
216               webcam                  NaN         NaN  

[217 rows x 6 columns]

In [ ]:
df.isna().sum()

item1      0
item2      0
item3      5
item4     86
item5    213
item6    215
dtype: int64

In [13]:
df.fillna("", axis=1, inplace=True)
df

item1                item2                item3  \
0          smartphone              earbuds     screen protector   
1              laptop                mouse             keyboard   
2          smartwatch      fitness tracker     wireless charger   
3                  TV             soundbar           HDMI cable   
4              tablet               stylus          tablet case   
..                ...                  ...                  ...   
212  desktop computer  mechanical keyboard         gaming mouse   
213        smartwatch      fitness tracker        charging dock   
214            laptop               webcam  external hard drive   
215            tablet     screen protector               stylus   
216  desktop computer       gaming monitor         RGB keyboard   

                   item4                item5       item6  
0             power bank            USB cable  phone case  
1           laptop stand  external hard drive              
2                                                          
3       streaming device                                   
4         screen cleaner                                   
..                   ...                  ...         ...  
212                                                        
213  wireless headphones                                   
214              USB hub                                   
215     portable charger                                   
216               webcam                                   

[217 rows x 6 columns]

In [ ]:
import numpy as np
len(set(np.ravel(df.values)))

93

In [20]:
items = df.apply(lambda x: [i for i in x if i != ''], axis=1).tolist()
items

[['smartphone',
  'earbuds',
  'screen protector',
  'power bank',
  'USB cable',
  'phone case'],
 ['laptop', 'mouse', 'keyboard', 'laptop stand', 'external hard drive'],
 ['smartwatch', 'fitness tracker', 'wireless charger'],
 ['TV', 'soundbar', 'HDMI cable', 'streaming device'],
 ['tablet', 'stylus', 'tablet case', 'screen cleaner'],
 ['smartphone', 'smartwatch', 'earbuds'],
 ['gaming console', 'game controller', 'HDMI cable', 'gaming headset'],
 ['laptop', 'wireless mouse', 'backpack', 'laptop sleeve'],
 ['smartphone', 'phone case', 'screen protector', 'USB cable'],
 ['smartphone', 'smartwatch', 'power bank'],
 ['laptop', 'external hard drive', 'USB hub'],
 ['TV', 'soundbar', 'HDMI cable'],
 ['smartphone', 'earbuds', 'power bank'],
 ['smartwatch', 'fitness tracker', 'charging dock'],
 ['gaming console', 'game controller', 'gaming headset'],
 ['laptop', 'keyboard', 'mouse', 'monitor'],
 ['tablet', 'stylus', 'tablet case'],
 ['smartphone', 'screen protector', 'phone case'],
 ['TV', '

### Problem 2 - Item sets
1. Print out the frequent item sets along with their support values also display the count of item sets.(**min_support=0.08**)
2. "Shocking Electronics" wants to replenish its stocks, help find the top 5 most popular(higher buying frequency) items/item_sets to replenish. **Explain and justify** the process followed to come to the conclusion.


In [22]:
from apyori import apriori

frequent_itemsets = apriori(items, min_support=0.08)

frequent_itemsets_list = list(frequent_itemsets)
for itemset in frequent_itemsets_list:
    print(f"Itemset: {itemset.items}, Support: {itemset.support}")


Itemset: frozenset({'HDMI cable'}), Support: 0.15668202764976957
Itemset: frozenset({'TV'}), Support: 0.1382488479262673
Itemset: frozenset({'earbuds'}), Support: 0.1152073732718894
Itemset: frozenset({'external hard drive'}), Support: 0.1382488479262673
Itemset: frozenset({'fitness tracker'}), Support: 0.11981566820276497
Itemset: frozenset({'gaming console'}), Support: 0.12903225806451613
Itemset: frozenset({'gaming headset'}), Support: 0.08755760368663594
Itemset: frozenset({'laptop'}), Support: 0.1889400921658986
Itemset: frozenset({'phone case'}), Support: 0.09216589861751152
Itemset: frozenset({'screen protector'}), Support: 0.15207373271889402
Itemset: frozenset({'smartphone'}), Support: 0.2350230414746544
Itemset: frozenset({'smartwatch'}), Support: 0.14285714285714285
Itemset: frozenset({'soundbar'}), Support: 0.12442396313364056
Itemset: frozenset({'streaming device'}), Support: 0.0967741935483871
Itemset: frozenset({'stylus'}), Support: 0.09216589861751152
Itemset: frozenset

In [23]:
item_support = {}
for itemset in frequent_itemsets_list:
    for item in itemset.items:
        item_support[item] = item_support.get(item, 0) + itemset.support

top_items = sorted(item_support.items(), key=lambda x: x[1], reverse=True)[:5]

print("Top 5 items to replenish:")
for item, support in top_items:
    print(f"Item: {item}, Support: {support}")


Top 5 items to replenish:
Item: smartphone, Support: 0.5483870967741935
Item: TV, Support: 0.45161290322580644
Item: soundbar, Support: 0.43317972350230416
Item: laptop, Support: 0.4147465437788018
Item: streaming device, Support: 0.3778801843317972


### Problem 3 - Association Rules
+ Items on the left side of the association rule are called : **Antecedent items** and the right side of the association rule are called : **Consequent Items**.
1. Print out the association rules along with their confidence and lift. (Analyse the output structure of apriori())
    + **(min_support=0.01, min_confidence = 0.08, min_lift=1.7, min_length=3)**
2. As the festive season is approaching, "Shocking Electronics" is considering to provide discounts and offers on some of their products. Help them identify the top 5 popular **pairs/sets** of items/item_sets bought, considering probability of consequent item being purchased when antecedent item is bought.
3. Also help them identify the top 5 popular **pairs/sets** of items/item_sets bought together, considering the popularity of consequent and antecedent items.
+ (Consequent and antecedent items together form the **pairs/sets** specified in the question)


In [26]:
print("Association Rules along with Confidence and Lift:\n")
for itemset in frequent_itemsets_list:
    for ordered_stat in itemset.ordered_statistics:
        if len(ordered_stat.items_base) > 0 and len(ordered_stat.items_add) > 0:
            antecedent = tuple(ordered_stat.items_base)
            consequent = tuple(ordered_stat.items_add)
            support = itemset.support
            confidence = ordered_stat.confidence
            lift = ordered_stat.lift

            print(f"Antecedent: {antecedent}, Consequent: {consequent}, Support: {support:.2f}, Confidence: {confidence:.2f}, Lift: {lift:.2f}")


Association Rules along with Confidence and Lift:

Antecedent: ('TV',), Consequent: ('soundbar',), Support: 0.12, Confidence: 0.90, Lift: 7.23
Antecedent: ('soundbar',), Consequent: ('TV',), Support: 0.12, Confidence: 1.00, Lift: 7.23
Antecedent: ('TV',), Consequent: ('streaming device',), Support: 0.10, Confidence: 0.70, Lift: 7.23
Antecedent: ('streaming device',), Consequent: ('TV',), Support: 0.10, Confidence: 1.00, Lift: 7.23
Antecedent: ('earbuds',), Consequent: ('smartphone',), Support: 0.10, Confidence: 0.88, Lift: 3.74
Antecedent: ('smartphone',), Consequent: ('earbuds',), Support: 0.10, Confidence: 0.43, Lift: 3.74
Antecedent: ('external hard drive',), Consequent: ('laptop',), Support: 0.13, Confidence: 0.97, Lift: 5.12
Antecedent: ('laptop',), Consequent: ('external hard drive',), Support: 0.13, Confidence: 0.71, Lift: 5.12
Antecedent: ('fitness tracker',), Consequent: ('smartwatch',), Support: 0.12, Confidence: 0.96, Lift: 6.73
Antecedent: ('smartwatch',), Consequent: ('fit

In [24]:
rules = []
for itemset in frequent_itemsets_list:
    for ordered_stat in itemset.ordered_statistics:
        if len(ordered_stat.items_base) > 0 and len(ordered_stat.items_add) > 0:
            rules.append({
                'antecedent': tuple(ordered_stat.items_base),
                'consequent': tuple(ordered_stat.items_add),
                'support': itemset.support,
                'confidence': ordered_stat.confidence,
                'lift': ordered_stat.lift
            })

rules_sorted_by_confidence = sorted(rules, key=lambda x: x['confidence'], reverse=True)[:5]
rules_sorted_by_lift = sorted(rules, key=lambda x: x['lift'], reverse=True)[:5]

print("Top 5 popular pairs/sets based on confidence:")
for rule in rules_sorted_by_confidence:
    print(f"Antecedent: {rule['antecedent']}, Consequent: {rule['consequent']}, Confidence: {rule['confidence']:.2f}, Lift: {rule['lift']:.2f}")

print("\nTop 5 popular pairs/sets based on lift:")
for rule in rules_sorted_by_lift:
    print(f"Antecedent: {rule['antecedent']}, Consequent: {rule['consequent']}, Confidence: {rule['confidence']:.2f}, Lift: {rule['lift']:.2f}")


Top 5 popular pairs/sets based on confidence:
Antecedent: ('soundbar',), Consequent: ('TV',), Confidence: 1.00, Lift: 7.23
Antecedent: ('streaming device',), Consequent: ('TV',), Confidence: 1.00, Lift: 7.23
Antecedent: ('wireless mouse',), Consequent: ('laptop',), Confidence: 1.00, Lift: 5.29
Antecedent: ('phone case',), Consequent: ('smartphone',), Confidence: 1.00, Lift: 4.25
Antecedent: ('streaming device', 'soundbar'), Consequent: ('TV',), Confidence: 1.00, Lift: 7.23

Top 5 popular pairs/sets based on lift:
Antecedent: ('stylus',), Consequent: ('tablet',), Confidence: 0.90, Lift: 10.28
Antecedent: ('tablet',), Consequent: ('stylus',), Confidence: 0.95, Lift: 10.28
Antecedent: ('streaming device',), Consequent: ('soundbar',), Confidence: 0.95, Lift: 7.65
Antecedent: ('streaming device',), Consequent: ('TV', 'soundbar'), Confidence: 0.95, Lift: 7.65
Antecedent: ('TV', 'streaming device'), Consequent: ('soundbar',), Confidence: 0.95, Lift: 7.65


### Problem 4 - Theory
Explain **Rule**, **Lift** and **Support**, taking examples from the results obtained from **your analysis**



1. **Rule**:
   - An **association rule** is a relationship between two or more items. It helps in discovering how items are associated with one another in a dataset. A rule is structured in the form `Antecedent → Consequent`, where the antecedent is the item(s) purchased first, and the consequent is the item(s) purchased after.
   
   **Example**:
   - **Rule**: `('soundbar',) → ('TV',)`
     - This rule indicates that customers who buy a soundbar also tend to buy a TV.

2. **Support**:
   - **Support** measures the frequency of occurrence of an itemset in the dataset. It is the ratio of transactions that contain the itemset to the total number of transactions. It helps determine how popular an itemset is.
   
   **Example**:
   - If the **support** for the rule `('soundbar',) → ('TV',)` is `0.15`, it means that 15% of all transactions include both a soundbar and a TV.

3. **Lift**:
   - **Lift** measures how much more likely the consequent is to be purchased when the antecedent is purchased, compared to the chance of purchasing the consequent independently. It indicates the strength of the association.
   
   - A **lift** greater than 1 indicates a strong positive association between the antecedent and the consequent.
   
   **Example**:
   - **Lift** for the rule `('soundbar',) → ('TV',)` is `7.23`, meaning that customers who buy a soundbar are 7.23 times more likely to buy a TV compared to customers who don't buy a soundbar.

